File: graphs/course_schedule_ii.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        adj = [[] for _ in range(numCourses)]
        in_degree = [0] * numCourses

        for c, pre in prerequisites:
            in_degree[c] += 1
            adj[pre].append(c)

        queue = deque([i for i in range(numCourses) if in_degree[i] == 0])
        result = []

        while queue:
            node = queue.popleft()
            result.append(node)
            for n in adj[node]:
                in_degree[n] -= 1
                if in_degree[n] == 0:
                    queue.append(n)

        return result if len(result) == numCourses else []

File: graphs/course_schedule_ii.rb
# @param {Integer} num_courses
# @param {Integer[][]} prerequisites
# @return {Integer[]}
def find_order(num_courses, prerequisites)
    adj = Array.new(num_courses) { [] }
    in_degree = Array.new(num_courses) { 0 }
    prerequisites.each do |c, pre|
        in_degree[c] += 1
        adj[pre] << c
    end
    queue = Queue.new

    (0...num_courses).each do |i|
        queue << i if in_degree[i] == 0
    end

    result = []

    until queue.empty? do
        node = queue.pop
        result << node
        adj[node].each do |n|
            in_degree[n] -= 1
            queue << n if in_degree[n] == 0
        end
    end 
    result.size == num_courses ? result : []
end